### PR TITLE
Add funcs that help us to compare and create trees

### PR DESCRIPTION
**Pull request by GoRunMaxim**
Upload Tree struct and its constructor. 
Upload funcs: Same, Walk and tests for them.

### CODE REVIEW

File: tree.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package tree
```
Comment (papahmsolo) [inline tree.go:1]: Package comment

File: tree.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+func insert(t *Tree, v int) *Tree {
+	if t == nil {
+		return &Tree{nil, v, nil}
+	}
+	if v < t.Value {
+		t.Left = insert(t.Left, v)
+	} else {
+		t.Right = insert(t.Right, v)
+	}
```
Comment (papahmsolo) [inline tree.go:54]: Do we need `else` here?

File: tree.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	go Walker(t2, c2)
+	var data1, data2 []int
+	for i := range c1 {
+		data1 = append(data1, i)
+	}
+	for i := range c2 {
+		data2 = append(data2, i)
+	}
+	sort.Ints(data1)
+	sort.Ints(data2)
```
Comment (papahmsolo) [inline tree.go:72]: Why sorting slices?

File: tree.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	c2 := make(chan int)
+	go Walker(t1, c1)
+	go Walker(t2, c2)
+	var data1, data2 []int
+	for i := range c1 {
+		data1 = append(data1, i)
+	}
+	for i := range c2 {
+		data2 = append(data2, i)
+	}
```
Comment (papahmsolo) [inline tree.go:70]: What's the point in sending tree values to the channel if you can just append to a slice while walking the tree?

**Review by papahmsolo — CHANGES_REQUESTED** [review]
This solution probably works fine for small trees. You can benchmark it now and then we'll compare it to a better solution. Hint: do we always need to walk over the whole trees to compare?  Or maybe we can sometimes tell trees are not equal in the middle of the walk?

File: tree.go (hunk 3 above)
Comment (GoRunMaxim) [reply tree.go:72]: Unnecessary work. Deleted

File: tree.go (hunk 4 above)
Comment (GoRunMaxim) [reply tree.go:70]: **Implement the Same function using Walk to determine whether t1 and t2 store the same values.**

Walk function reads tree values and send them to channel, then we read this values and append to slice. Don't we need to do this?

File: tree.go (hunk 2 above)
Comment (GoRunMaxim) [reply tree.go:54]: I suppose we do. Or what is the better way to insert the node?

File: tree.go (hunk 1 above)
Comment (GoRunMaxim) [reply tree.go:3]: Fixed

File: tree.go (hunk 4 above)
Comment (GoRunMaxim) [reply tree.go]: Fixed

File: tree.go (hunk 2 above)
Comment (papahmsolo) [reply tree.go:45]: It always falls into the `else` category, so if you just delete `else` it wouldn't change anything

File: tree.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	walk = func(t *Tree) {
+		if t == nil {
+			return
+		}
+		walk(t.Left)
+		ch <- t.Value
+		walk(t.Right)
+	}
+	walk(t)
+}
```
Comment (papahmsolo) [inline tree.go:26]: That's a nice way to walk a tree of arbitrary length! I found your previous version with `Walk` and `Walker` a little simpler, to be honest, but this works just fine. Pick the one you like better.

File: tree_test.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+			cValue, ok := <-ch
+			if !ok {
+				t.Errorf("Cannot read from close channel")
+			}
+			if tValue != cValue {
+				t.Errorf("Data mismatch: Want: %v, got %v", tValue, cValue)
+			}
+		}
+	}
+}
```
Comment (papahmsolo) [inline tree_test.go:38]: That is a nice test! As a suggestion, I'd return the number of actual nodes sent to the channel if it was closed earlier than expected

File: tree_test.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+func TestSame(t *testing.T) {
+	tests := []struct {
+		data1 []int
+		data2 []int
+		equal bool
```
Comment (papahmsolo) [inline tree_test.go:44]: I'd call them `nodes` rather than `data`

File: tree_test.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+			data2: []int{1, 2, 3, 4, 5},
+			equal: false,
+		},
+	}
+	for _, test := range tests {
+		t1 := New(test.data1...)
+		t2 := New(test.data2...)
+		if Same(t1, t2) != test.equal {
+			t.Errorf("Should be same: t1: %v, t2 %v", t1, t2)
+		}
```
Comment (papahmsolo) [inline tree_test.go:72]: Try to fail this test deliberately, you'll find the error message does not provide many details about the problem. What about printing the contents of these trees?

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Looks very good!
I've made some suggestions to improve your code.

File: tree.go (hunk 5 above)
Comment (GoRunMaxim) [reply tree.go:26]: **It always falls into the else category, so if you just delete else it wouldn't change anything**

That doesn't work like this, we do need else statement. Try it by urself. 

**Pick the one you like better.**

Let's just leave only one long function.

File: tree_test.go (hunk 7 above)
Comment (GoRunMaxim) [reply tree_test.go:44]: It's  actually not nodes by themselves, but the value that this nodes has. But I will change this.

File: tree.go (hunk 5 above)
Comment (papahmsolo) [reply tree.go:26]: Oh, I see the problem, missed the `return` statement. I just do it this way

File: tree_test.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	for _, test := range tests {
+		ch := make(chan int)
+		tree := New(test.data...)
+		sort.Ints(test.data) // Walk function returns ascending values
+		go Walk(tree, ch)
+		for tNumber, tValue := range test.data {
+			cValue, ok := <-ch
+			if !ok {
+				t.Errorf("Cannot read from close channel, nodes number is %v", tNumber)
+			}
```
Comment (papahmsolo) [inline tree_test.go:33]: You can specify in the error message what you expected and what you actually got

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Almost there

**Review by papahmsolo — CHANGES_REQUESTED** [review]
Well done!

**Review by papahmsolo — APPROVED** [review]
Well done!

**Pull request merged**